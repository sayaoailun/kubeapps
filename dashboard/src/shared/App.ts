import { axiosWithAuth } from "./AxiosInstance";
import { hapi } from "./hapi/release";
import { IAppOverview, IChartVersion } from "./types";

export const TILLER_PROXY_ROOT_URL = "api/tiller-deploy/v1";

export class App {
  public static getResourceURL(namespace?: string, name?: string, query?: string) {
    let url = TILLER_PROXY_ROOT_URL;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += "/releases";
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static async create(
    releaseName: string,
    namespace: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace);
    const { data } = await axiosWithAuth.post(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async upgrade(
    releaseName: string,
    namespace: string,
    chartNamespace: string,
    chartVersion: IChartVersion,
    values?: string,
  ) {
    const chartAttrs = chartVersion.relationships.chart.data;
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName: chartAttrs.repo.name,
      appRepositoryResourceNamespace: chartNamespace,
      chartName: chartAttrs.name,
      releaseName,
      values,
      version: chartVersion.attributes.version,
    });
    return data;
  }

  public static async rollback(releaseName: string, namespace: string, revision: number) {
    const endpoint = App.getResourceURL(namespace, releaseName);
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(releaseName: string, namespace: string, purge: boolean) {
    let purgeQuery;
    if (purge) {
      purgeQuery = "purge=true";
    }
    const { data } = await axiosWithAuth.delete(
      App.getResourceURL(namespace, releaseName, purgeQuery),
    );
    return data;
  }

  public static async listApps(namespace?: string, allStatuses?: boolean) {
    let query;
    if (allStatuses) {
      query = "statuses=all";
    }
    const { data } = await axiosWithAuth.get<{ data: IAppOverview[] }>(
      App.getResourceURL(namespace, undefined, query),
    );
    return data.data;
  }

  public static async getRelease(namespace: string, name: string) {
    const { data } = await axiosWithAuth.get<{ data: hapi.release.Release }>(
      this.getResourceURL(namespace, name),
    );
    return data.data;
  }
}
