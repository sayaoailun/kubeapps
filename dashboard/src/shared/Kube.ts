import { Auth } from "./Auth";
import { axiosWithAuth } from "./AxiosInstance";
import { ResourceKindsWithAPIVersions } from "./ResourceAPIVersion";
import { ResourceKind, ResourceKindsWithPlurals } from "./ResourceKinds";
import { IK8sList, IResource } from "./types";

export const APIBase = "api/kube";
export let WebSocketAPIBase: string;
if (location.protocol === "https:") {
  WebSocketAPIBase = `wss://${window.location.host}${window.location.pathname}`;
} else {
  WebSocketAPIBase = `ws://${window.location.host}${window.location.pathname}`;
}

// Kube is a lower-level class for interacting with the Kubernetes API. Use
// ResourceRef to interact with a single API resource rather than using Kube
// directly.
export class Kube {
  public static getResourceURL(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let url = `${APIBase}/${apiVersion === "v1" ? "api/v1" : `apis/${apiVersion}`}`;
    if (namespace) {
      url += `/namespaces/${namespace}`;
    }
    url += `/${resource}`;
    if (name) {
      url += `/${name}`;
    }
    if (query) {
      url += `?${query}`;
    }
    return url;
  }

  public static watchResourceURL(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    let url = this.getResourceURL(apiVersion, resource, namespace);
    url = `${WebSocketAPIBase}${url}?watch=true`;
    if (name) {
      url += `&fieldSelector=metadata.name%3D${name}`;
    }
    if (query) {
      url += `&${query}`;
    }
    return url;
  }

  public static async getResource(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    const { data } = await axiosWithAuth.get<IResource | IK8sList<IResource, {}>>(
      this.getResourceURL(apiVersion, resource, namespace, name, query),
    );
    return data;
  }

  // Opens and returns a WebSocket for the requested resource. Note: it is
  // important that this socket be properly closed when no longer needed. The
  // returned WebSocket can be attached to an event listener to read data from
  // the socket.
  public static watchResource(
    apiVersion: string,
    resource: string,
    namespace?: string,
    name?: string,
    query?: string,
  ) {
    return new WebSocket(
      this.watchResourceURL(apiVersion, resource, namespace, name, query),
      Auth.wsProtocols(),
    );
  }

  // Gets the plural form of the resource Kind for use in the resource path
  public static resourcePlural(kind: ResourceKind) {
    return ResourceKindsWithPlurals[kind];
  }

  // Gets the apiVersion of the resource Kind for use in the resource path
  public static resourceAPIVersion(kind: ResourceKind) {
    return ResourceKindsWithAPIVersions[kind];
  }
}
