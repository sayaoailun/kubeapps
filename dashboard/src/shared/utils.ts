export function escapeRegExp(str: string) {
  return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&");
}

export function getValueFromEvent(e: React.FormEvent<HTMLInputElement>) {
  let value: any = e.currentTarget.value;
  switch (e.currentTarget.type) {
    case "checkbox":
      // value is a boolean
      value = e.currentTarget.checked;
      break;
    case "number":
      // value is a number
      value = e.currentTarget.valueAsNumber;
      break;
  }
  return value;
}

export async function wait(ms: number = 1) {
  await new Promise(resolve => setTimeout(() => resolve(), ms));
}
