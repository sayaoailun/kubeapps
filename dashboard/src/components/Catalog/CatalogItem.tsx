import * as React from "react";
import { Link } from "react-router-dom";

import helmIcon from "../../icons/helm.svg";
import operatorIcon from "../../icons/operator-framework.svg";
import placeholder from "../../placeholder.png";
import { IRepo } from "../../shared/types";
import InfoCard from "../InfoCard";

import "./CatalogItem.css";

interface ICatalogItemProps {
  item: ICatalogItem;
}

export interface ICatalogItem {
  id: string;
  name: string;
  version: string;
  description: string;
  type: "chart" | "operator";
  namespace: string;
  icon?: string;
  repo?: IRepo;
  csv?: string;
}

// 3 lines description max
const MAX_DESC_LENGTH = 90;

function trimDescription(desc: string): string {
  if (desc.length > MAX_DESC_LENGTH) {
    // Trim to the last word under the max length
    return desc.substr(0, desc.lastIndexOf(" ", MAX_DESC_LENGTH)).concat("...");
  }
  return desc;
}

const CatalogItem: React.SFC<ICatalogItemProps> = props => {
  const { item } = props;
  const { icon, name, repo, version, description, type, namespace, id, csv } = item;
  const iconSrc = icon || placeholder;
  let link;
  let tag1;
  let subIcon;
  if (type === "chart") {
    tag1 = (
      <Link
        className="ListItem__content__info_tag_link"
        to={`/ns/${namespace}/catalog/${repo?.name}`}
      >
        {repo?.name}
      </Link>
    );
    const chartsSegment = namespace !== repo?.namespace ? "global-charts" : "charts";
    link = `/ns/${namespace}/${chartsSegment}/${repo?.name}/${name}`;
    subIcon = helmIcon;
  } else {
    // Cosmetic change, remove the version from the csv name
    const csvName = csv?.split(".v")[0];
    tag1 = <span>{csvName}</span>;
    link = `/ns/${namespace}/operators-instances/new/${csv}/${id}`;
    subIcon = operatorIcon;
  }
  const descriptionC = (
    <div className="ListItem__content__description">{trimDescription(description)}</div>
  );
  return (
    <InfoCard
      key={id}
      title={name}
      link={link}
      info={version || "-"}
      icon={iconSrc}
      description={descriptionC}
      tag1Content={tag1}
      tag1Class={repo ? repo.name : ""}
      subIcon={subIcon}
    />
  );
};

export default CatalogItem;
