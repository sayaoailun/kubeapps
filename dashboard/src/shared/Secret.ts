import { axiosWithAuth } from "./AxiosInstance";
import { APIBase } from "./Kube";
import { IK8sList, IOwnerReference, ISecret } from "./types";

export default class Secret {
  public static async create(
    name: string,
    secrets: { [s: string]: string },
    owner: IOwnerReference | undefined,
    namespace: string,
  ) {
    const url = Secret.getLink(namespace);
    const { data } = await axiosWithAuth.post<ISecret>(url, {
      apiVersion: "v1",
      data: secrets,
      kind: "Secret",
      metadata: {
        name,
        ownerReferences: [owner],
      },
      type: "Opaque",
    });
    return data;
  }

  public static async delete(name: string, namespace: string) {
    const url = this.getLink(namespace, name);
    return axiosWithAuth.delete(url);
  }

  public static async get(name: string, namespace: string) {
    const url = this.getLink(namespace, name);
    const { data } = await axiosWithAuth.get<ISecret>(url);
    return data;
  }

  public static async list(namespace: string) {
    const url = Secret.getLink(namespace);
    const { data } = await axiosWithAuth.get<IK8sList<ISecret, {}>>(url);
    return data;
  }

  public static async createPullSecret(
    name: string,
    user: string,
    password: string,
    email: string,
    server: string,
    namespace: string,
  ) {
    const url = Secret.getLink(namespace);
    const dockercfg = {
      auths: {
        [server]: {
          username: user,
          password,
          email,
          auth: btoa(`${user}:${password}`),
        },
      },
    };
    const { data } = await axiosWithAuth.post<ISecret>(url, {
      apiVersion: "v1",
      stringData: {
        ".dockerconfigjson": JSON.stringify(dockercfg),
      },
      kind: "Secret",
      metadata: {
        name,
      },
      type: "kubernetes.io/dockerconfigjson",
    });
    return data;
  }

  private static getLink(namespace: string, name?: string): string {
    return `${APIBase}/api/v1/namespaces/${namespace}/secrets${name ? `/${name}` : ""}`;
  }
}
