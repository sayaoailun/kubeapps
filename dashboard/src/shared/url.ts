import { APIBase } from "./Kube";
import { IServiceBroker } from "./ServiceCatalog";
import { IChartVersion } from "./types";

export const app = {
  charts: {
    version: (cv: IChartVersion) =>
      `/charts/${cv.relationships.chart.data.repo.name}/${cv.relationships.chart.data.name}/versions/${cv.attributes.version}`,
  },
  operators: {
    view: (namespace: string, name: string) => `/ns/${namespace}/operators/${name}`,
    list: (namespace: string) => `/ns/${namespace}/operators`,
    new: (namespace: string, name: string) => `/ns/${namespace}/operators/new/${name}`,
  },
  operatorInstances: {
    update: (namespace: string, csvName: string, crdName: string, instanceName: string) =>
      `/ns/${namespace}/operators-instances/${csvName}/${crdName}/${instanceName}/update`,
  },
};

function withNS(namespace: string) {
  return namespace === "_all" ? "" : `namespaces/${namespace}/`;
}

export const backend = {
  apprepositories: {
    base: (namespace: string) => `api/v1/namespaces/${namespace}/apprepositories`,
    create: (namespace: string) => backend.apprepositories.base(namespace),
    validate: () => `${backend.apprepositories.base("kubeapps")}/validate`,
    delete: (name: string, namespace: string) =>
      `${backend.apprepositories.base(namespace)}/${name}`,
    update: (namespace: string, name: string) =>
      `${backend.apprepositories.base(namespace)}/${name}`,
  },
  namespaces: {
    base: "api/v1/namespaces",
    list: () => `${backend.namespaces.base}`,
  },
};

export const api = {
  apprepostories: {
    base: `${APIBase}/apis/kubeapps.com/v1alpha1`,
    create: (namespace: string) =>
      `${api.apprepostories.base}/namespaces/${namespace}/apprepositories`,
  },

  charts: {
    base: "api/assetsvc/v1",
    get: (namespace: string, id: string) => `${api.charts.base}/ns/${namespace}/charts/${id}`,
    getReadme: (namespace: string, id: string, version: string) =>
      `${api.charts.base}/ns/${namespace}/assets/${id}/versions/${encodeURIComponent(
        version,
      )}/README.md`,
    getValues: (namespace: string, id: string, version: string) =>
      `${api.charts.base}/ns/${namespace}/assets/${id}/versions/${encodeURIComponent(
        version,
      )}/values.yaml`,
    getSchema: (namespace: string, id: string, version: string) =>
      `${api.charts.base}/ns/${namespace}/assets/${id}/versions/${encodeURIComponent(
        version,
      )}/values.schema.json`,
    getVersion: (namespace: string, id: string, version: string) =>
      `${api.charts.base}/ns/${namespace}/charts/${id}/versions/${encodeURIComponent(version)}`,
    list: (namespace: string, repo?: string) =>
      `${api.charts.base}/ns/${namespace}/charts${repo ? `/${repo}` : ""}`,
    listVersions: (namespace: string, id: string) => `${api.charts.get(namespace, id)}/versions`,
  },

  serviceinstances: {
    base: `${APIBase}/apis/servicecatalog.k8s.io/v1beta1`,
    create: (namespace: string) =>
      `${api.serviceinstances.base}/namespaces/${namespace}/serviceinstances`,
  },

  clusterservicebrokers: {
    base: `${APIBase}/apis/servicecatalog.k8s.io/v1beta1`,
    sync: (broker: IServiceBroker) =>
      `${api.clusterservicebrokers.base}/clusterservicebrokers/${broker.metadata.name}`,
  },

  operators: {
    operators: (namespace: string) =>
      `${APIBase}/apis/packages.operators.coreos.com/v1/${withNS(namespace)}packagemanifests`,
    operator: (namespace: string, name: string) =>
      `${APIBase}/apis/packages.operators.coreos.com/v1/namespaces/${namespace}/packagemanifests/${name}`,
    clusterServiceVersions: (namespace: string) =>
      `${APIBase}/apis/operators.coreos.com/v1alpha1/${withNS(namespace)}clusterserviceversions`,
    clusterServiceVersion: (namespace: string, name: string) =>
      `${APIBase}/apis/operators.coreos.com/v1alpha1/namespaces/${namespace}/clusterserviceversions/${name}`,
    operatorIcon: (namespace: string, name: string) =>
      `api/v1/namespaces/${namespace}/operator/${name}/logo`,
    resources: (namespace: string, apiVersion: string, resource: string) =>
      `${APIBase}/apis/${apiVersion}/${withNS(namespace)}${resource}`,
    resource: (namespace: string, apiVersion: string, resource: string, name: string) =>
      `${APIBase}/apis/${apiVersion}/namespaces/${namespace}/${resource}/${name}`,
    operatorGroups: (namespace: string) =>
      `${APIBase}/apis/operators.coreos.com/v1/namespaces/${namespace}/operatorgroups`,
    subscription: (namespace: string, name: string) =>
      `${APIBase}/apis/operators.coreos.com/v1alpha1/namespaces/${namespace}/subscriptions/${name}`,
  },
};
