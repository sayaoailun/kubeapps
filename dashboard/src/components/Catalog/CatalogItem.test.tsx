import { shallow } from "enzyme";
import context from "jest-plugin-context";
import { cloneDeep } from "lodash";
import * as React from "react";

import { IRepo } from "../../shared/types";
import { CardIcon } from "../Card";
import InfoCard from "../InfoCard";
import CatalogItem, { ICatalogItem } from "./CatalogItem";

jest.mock("../../placeholder.png", () => "placeholder.png");

const defaultItem = {
  id: "foo1",
  name: "foo",
  version: "1.0.0",
  description: "",
  type: "chart",
  repo: {
    name: "repo-name",
    namespace: "repo-namespace",
  } as IRepo,
  namespace: "repo-namespace",
  icon: "icon.png",
} as ICatalogItem;

it("should render a chart item in a namespace", () => {
  const wrapper = shallow(<CatalogItem item={defaultItem} />);
  expect(wrapper).toMatchSnapshot();
});

it("should render a global chart item in a namespace", () => {
  const globalItem = {
    ...defaultItem,
    repo: {
      name: "repo-name",
      namespace: "kubeapps",
    } as IRepo,
  };
  const wrapper = shallow(<CatalogItem item={globalItem} />);
  expect(wrapper).toMatchSnapshot();
});

it("should use the default placeholder for the icon if it doesn't exist", () => {
  const chartWithoutIcon = cloneDeep(defaultItem);
  chartWithoutIcon.icon = undefined;
  const wrapper = shallow(<CatalogItem item={chartWithoutIcon} />);
  // Importing an image returns "undefined"
  expect(
    wrapper
      .find(InfoCard)
      .shallow()
      .find(CardIcon)
      .prop("src"),
  ).toBe(undefined);
});

it("should place a dash if the version is not avaliable", () => {
  const chartWithoutVersion = cloneDeep(defaultItem);
  chartWithoutVersion.version = "";
  const wrapper = shallow(<CatalogItem item={chartWithoutVersion} />);
  expect(
    wrapper
      .find(InfoCard)
      .shallow()
      .find(".type-color-light-blue")
      .text(),
  ).toBe("-");
});

it("show the chart description", () => {
  const chartWithDescription = cloneDeep(defaultItem);
  chartWithDescription.description = "This is a description";
  const wrapper = shallow(<CatalogItem item={chartWithDescription} />);
  expect(
    wrapper
      .find(InfoCard)
      .shallow()
      .find(".ListItem__content__description")
      .text(),
  ).toBe(chartWithDescription.description);
});

context("when the description is too long", () => {
  it("trims the description", () => {
    const chartWithDescription = cloneDeep(defaultItem);
    chartWithDescription.description =
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum ultrices velit leo, quis pharetra mi vestibulum quis.";
    const wrapper = shallow(<CatalogItem item={chartWithDescription} />);
    expect(
      wrapper
        .find(InfoCard)
        .shallow()
        .find(".ListItem__content__description")
        .text(),
    ).toMatch(/\.\.\.$/);
  });
});

context("when the item is a catalog", () => {
  const catalogItem = {
    ...defaultItem,
    csv: "foo-cluster",
    type: "operator",
  } as ICatalogItem;

  it("shows the proper tag", () => {
    const wrapper = shallow(<CatalogItem item={catalogItem} />);
    expect((wrapper.find(InfoCard).prop("tag1Content") as JSX.Element).props.children).toEqual(
      "foo-cluster",
    );
  });

  it("has the proper link", () => {
    const wrapper = shallow(<CatalogItem item={catalogItem} />);
    expect(wrapper.find(InfoCard).prop("link")).toEqual(
      `/ns/${defaultItem.namespace}/operators-instances/new/foo-cluster/foo1`,
    );
  });
});
