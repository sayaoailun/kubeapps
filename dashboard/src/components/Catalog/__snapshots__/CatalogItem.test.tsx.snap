// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a chart item in a namespace 1`] = `
<InfoCard
  description={
    <div
      className="ListItem__content__description"
    />
  }
  icon="icon.png"
  info="1.0.0"
  key="foo1"
  link="/ns/repo-namespace/charts/repo-name/foo"
  tag1Class="repo-name"
  tag1Content={
    <Unknown
      className="ListItem__content__info_tag_link"
      to="/ns/repo-namespace/catalog/repo-name"
    >
      repo-name
    </Unknown>
  }
  title="foo"
/>
`;

exports[`should render a global chart item in a namespace 1`] = `
<InfoCard
  description={
    <div
      className="ListItem__content__description"
    />
  }
  icon="icon.png"
  info="1.0.0"
  key="foo1"
  link="/ns/repo-namespace/global-charts/repo-name/foo"
  tag1Class="repo-name"
  tag1Content={
    <Unknown
      className="ListItem__content__info_tag_link"
      to="/ns/repo-namespace/catalog/repo-name"
    >
      repo-name
    </Unknown>
  }
  title="foo"
/>
`;
